### PR TITLE
Reorg crates

### PR DESCRIPTION
**Pull request by koushiro**
~~waiting for #192 to be merged firstly.~~

* reorg the crates: 
  * move `ReadOnlyDB` trait into `substrate-archive-backend`
  * move `types`, `models` and `util` modules into `substrate-archive`
  * merge `block_exec` and `tasks`, `sql_block_builder` and `models` 
  * remove `substrate-archive-common`
* ~~improve error handling (#194)~~

### CODE REVIEW

**Comment by insipx** [comment]
Why get rid of `common`?

Ideally I would like to merge this after #134 , since it would be a PITA to merge all the reorg changes into another big PR

**Comment by koushiro** [comment]
> Why get rid of `common`?
> 
> Ideally I would like to merge after #134 , since it would be a PITA to merge all the reorg changes into another big PR

IMO, `substrate-archive-common` is a boring crate, it just extracts some **non-public** parts from the two crates of `substrate-archive-backend` and `subtrate-archive` (like `ReadOnlyDB` trait used by backend, `types`, `models` and `logger` used by archive), and also messy error types.

I don't mind one more rebase modification after merging #134

**Comment by koushiro** [comment]
@insipx PTAL

**Review by insipx — APPROVED** [review]

**Comment by insipx** [comment]
Overall good changes, condenses code where it was 'hidden' otherwise

**Comment by koushiro** [comment]
@dvdplm @niklasad1 PTAL

File: substrate-archive/src/wasm_tracing.rs
Comment (dvdplm) [inline substrate-archive/src/wasm_tracing.rs:167]: Is this about silencing the complaint on line 175 (suggestion: "use t.level")? If so that seems to be a clippy bug and not something worth silencing like this imo.

File: substrate-archive/src/archive.rs
Comment (dvdplm) [inline substrate-archive/src/archive.rs:37]:
```suggestion
	migrations, 
	util,
```

File: substrate-archive/src/logger.rs
Comment (dvdplm) [inline substrate-archive/src/logger.rs:1]: All files need license headers. Please revert.

File: substrate-archive/src/logger.rs
Comment (dvdplm) [inline substrate-archive/src/logger.rs:17]: All modules should have module documentation; if anything improve and extend, but please do not remove any existing docs.

**Review by dvdplm — CHANGES_REQUESTED** [review]
Not sure I see the value of doing this. It's @insipx's call but I'm not convinced.

File: substrate-archive/src/archive.rs
Comment (koushiro) [reply substrate-archive/src/archive.rs:37]: it's result of command `cargo fmt`

File: substrate-archive/src/wasm_tracing.rs
Comment (koushiro) [reply substrate-archive/src/wasm_tracing.rs:167]: Yes, it seems to be a clippy bug (rust-lang/rust-clippy#6722) of rust 1.50, which needs to be fixed by the upstream

Comment (koushiro) [reply substrate-archive/src/wasm_tracing.rs:167]: @dvdplm PTAL

**Comment by koushiro** [comment]
I will improve the running configuration based on this PR, so hope the PR can be reviewed and merged ASAP @insipx @dvdplm

**Review by dvdplm — APPROVED** [review]

File: substrate-archive-backend/src/database.rs
Comment (niklasad1) [inline substrate-archive-backend/src/database.rs:39]:
```suggestion
	/// Open database as read-only
```

File: substrate-archive/src/error.rs
Comment (niklasad1) [inline substrate-archive/src/error.rs:24]: I fine with removing the comment or fixing the comment

```suggestion
	// Rust std io error
```

**Comment by insipx** [comment]
I think this PR is good, and I greatly appreciate your continued contributions.

I think the reason for the delay of this PR is that many of the included changes are stylistic/aesthetic/organizational. While often they make code more readable, substrate-archives codebase is nonetheless under flux and will probably change significantly in the future anyway. So it's sort of sisyphean to try and make the structure perfect now, when there are things like type-decoding/substrate runtime changes/paritydb/pluggable backends in the pipeline that will change the structure anyways. These features also move the project forward toward it's end-goal rather than stagnating on organization. I know it can be annoying to work in a codebase where things aren't where they 'should' be, or where it seems inefficient, but I don't think now is the right time to focus on such changes.


TL;DR I think we should stick with PR's that move the project forward toward it's goal of 'Generic Indexer pluggable for all Substrate Chains' rather than stagnate on organizing the codebase 'correctly'.

In that vein I think PR #227 (I still have to read through all of it, though) is in a better direction, those changes make it clearer to the end user, make config options easier to use and add features which were lacking before (and close two open issues!).


I'll approve this PR but will be more weary towards more stylistic changes in the future

**Comment by koushiro** [comment]
> TL;DR I think we should stick with PR's that move the project forward toward it's goal of 'Generic Indexer pluggable for all Substrate Chains' rather than stagnate on organizing the codebase 'correctly'.

@insipx  Thank you for your detailed reply, I totally agree with it, this will also be what I expected to see

**Pull request merged**
